src/device-manager/cocoa/NLWdmClient.mm

```diff
@@ -225,6 +227,48 @@ static void handleWdmClientError(
     [client dispatchAsyncDefaultFailureBlock:error];
 }
 
+static void onWdmClientFlushUpdateComplete(
+    void * appState, void * appReqState, uint16_t pathCount, nl::Weave::DeviceManager::WdmClientFlushUpdateStatus * statusResults)
+{
+    WDM_LOG_DEBUG(@"onWdmClientFlushUpdateComplete");
+
+    NLWdmClient * client = (__bridge NLWdmClient *) appReqState;
+
+    NSMutableArray * statusResultsList = [NSMutableArray arrayWithCapacity:pathCount];
+    NLGenericTraitUpdatableDataSink * dataSink = nil;
+    WDM_LOG_DEBUG(@"Failed path Count = %u\n", pathCount);
+
+    for (uint32_t i = 0; i < pathCount; i++) {
+        dataSink = [client getDataSink:(long long) statusResults[i].mpDataSink];
+        if (dataSink == nil) {
+            WDM_LOG_DEBUG(@"trait %d does not exist", i);
+        } else {
+            if (statusResults[i].mErrorCode == WEAVE_ERROR_STATUS_REPORT_RECEIVED) {
+                NLWdmClientFlushUpdateDeviceStatusError * statusError = [[NLWdmClientFlushUpdateDeviceStatusError alloc]
+                    initWithProfileId:statusResults[i].mDevStatus.StatusProfileId
+                           statusCode:statusResults[i].mDevStatus.StatusCode
+                            errorCode:statusResults[i].mDevStatus.SystemErrorCode
+                         statusReport:[client statusReportToString:statusResults[i].mDevStatus.StatusProfileId
+                                                        statusCode:statusResults[i].mDevStatus.StatusCode]
+                                 path:[NSString stringWithUTF8String:statusResults[i].mpPath]
+                             dataSink:dataSink];
+
+                [statusResultsList addObject:statusError];
+            } else {
+                NLWdmClientFlushUpdateError * weaveError = [[NLWdmClientFlushUpdateError alloc]
+                    initWithWeaveError:statusResults[i].mErrorCode
+                                report:[NSString stringWithUTF8String:nl::ErrorStr(statusResults[i].mErrorCode)]
+                                  path:[NSString stringWithUTF8String:statusResults[i].mpPath]
+                              dataSink:dataSink];
+                [statusResultsList addObject:weaveError];
+            }
+        }
+    }
+
+    NSLog(@"statusResultsList is: %@", statusResultsList);
+    [client dispatchAsyncCompletionBlockWithError:statusResultsList];
```

Comment:
Thanks for the suggestions.

statusResult would be filled with error results when there are failed path, which would not have any success results.  If statusResult is empty array, it means no path error for flush update operation.

OnError is used to notify the protocol layer error.
OnComplete is used to notify the results when the operation is completed and there is no protocol layer error.
For flush update operations, for example, one particular path cannot be modified because of authentication issue or others, and other path has been handled correctly, the flush operation is still overally seen as successful operation, which has been designed/implemented from c++ wdm update protocol for the long time, therefore we should use onFlushUpdateComplete with particular path. This has been applied from c++ protocol layer, JAVA, python, and objective C. In this situation, you get onFlushUpdateComplete, and you read different errors, and can delete the local cached data for that particular path, you can pop or suppress the notification to UI, the overall flush operation is still seen as valid, the flow can go forward.

From objective c application layer, you can do something as below.

    [wdmClient flushUpdate:^(id owner, id result1)
     {
         flushDone = YES;
         NSLog(@"error%@",result1);
 
         for (id object in result1)
         {
             NLWdmClientFlushUpdateDeviceStatusError *wdmClientDeviceStatus;
             NLWdmClientFlushUpdateError *wdmClientError;
             if ([object isKindOfClass:[NLWdmClientFlushUpdateDeviceStatusError class]]) {
                 // You have a mutable array!
                 wdmClientDeviceStatus = object;
                 NSLog(@"path%@",wdmClientDeviceStatus.path);
                 //[wdmClientDeviceStatus.dataSink deleteData:wdmClientDeviceStatus.path];
                 
             } else {
                 wdmClientError = object;
                 NSLog(@"path%@",wdmClientError.path);
                 [wdmClientError.dataSink deleteData:wdmClientError.path];
                 // You don't have a mutable array!
             }
 

         }
     }
                   failure:^(id owner, NSError *error)
     {
         flushDone = YES;
         gotError = YES;
         NSLog(@"Flush Update Error:%@", error);
         NSDictionary* params = [NSDictionary dictionaryWithObjectsAndKeys:@"Flush Update Error", @"Title", error, @"Error", nil];
         [self performSelectorOnMainThread:@selector(createErrorSheet:) withObject:params waitUntilDone:YES];
     }];

Thanks

Reply:
Okay, I see what you mean. From the client perspective, it's not clear to me that the distinction between a protocol layer error and an application/operation layer error is so obvious – either way, the client is going to need to determine how to handle the failure in either of those cases, which would often have a decent amount of overlap. But it sounds like it's just an OpenWeave standard that I'm not familiar with and is outside the scope of this CL.

In that case, just a couple small nits:
1. I think it would be quite helpful to have a documentation comment for |flushUpdate| in the header explaining the expected values/types that will be passed to the `WdmClientCompletionBlock`(and the meaning, ie. that if the success block is called with an empty array, that indicates that all paths were flushed successfully)
2. Renaming `-[NLWdmClient dispatchAsyncCompletionBlockWithError:]` so it's a little more clear why it's not calling the failure block – maybe `-[NLWdmClient dispatchAsyncCompletionBlockWithResults:]`? WDYT?

Reply:
Thanks, just address the comments:
a. add detailed descriptions on FlushUpdate API in header.
b. renaming toward dispatchAsyncCompletionBlockWithResults.

Thanks!